packages/styled-components/src/types.ts

```diff
@@ -230,15 +237,25 @@ export interface IInlineStyle<Props extends object> {
   generateStyleObject(executionContext: ExecutionContext & Props): object;
 }
 
-export type StyledObject<Props extends object> = CSS.Properties<number | (string & {})> & {
-  [key: string]:
-    | string
-    | number
-    | StyleFunction<Props>
-    | StyledObject<Props>
-    | RuleSet<any>
-    | undefined;
-};
+export type CSSProperties = CSS.Properties<number | (string & {})>;
+
+export type CSSPseudos = { [K in CSS.Pseudos]?: CSSObject };
+
+export type CSSKeyframes = object & { [key: string]: CSSObject };
+
+export type CSSObject = StyledObject<any>;
+
+export type StyledObject<Props extends object> = CSSProperties &
+  CSSPseudos & {
+    [key: string]:
+      | StyledObject<any> // StyledObject<any> instead of CSSObject. Because writing CSSObject directly results in a circularly references.
+      | string
+      | number
+      | StyleFunction<Props>
+      | StyledObject<Props>
+      | RuleSet<any>
+      | undefined;
+  };
```

Comment:
At lines 251 and 255 we have the same type but with different generics being passed. Is it ok?

Reply:
Ahh, that seems incorrect.

Reply:
@pedroSoaresll 
Thank you for your comment. It is fixed here.
https://github.com/styled-components/styled-components/pull/4140#discussion_r1302280083

Reply:
Furthermore, I found a duplicate type definition, so I'll fix it here. https://github.com/styled-components/styled-components/pull/4200